crates/exex/src/hints.rs

```diff
@@ -128,3 +156,119 @@ pub fn print_tx_hint() -> Hint {
+                .map_err(HintError::Memory);
+
+            // Insert the chain ID of the transaction, defaulting to 0 if not available.
+            let _ = vm
+                .insert_value(
+                    chain_id_addr,
+                    Felt252::from(transaction.chain_id().unwrap_or_default()),
+                )
+                .map_err(HintError::Memory);
+
+            // Insert the low and high 16 bytes of the previous block's RANDAO (mix hash) into the
+            // VM.
+            let _ = vm
+                .insert_value(
+                    prev_randao_low_addr,
+                    Felt252::from_bytes_be_slice(&block.mix_hash.0[16..]),
+                )
+                .map_err(HintError::Memory);
+            let _ = vm
+                .insert_value(
+                    prev_randao_high_addr,
+                    Felt252::from_bytes_be_slice(&block.mix_hash.0[0..16]),
+                )
+                .map_err(HintError::Memory);
+
+            // Insert the block number into the VM.
+            let _ = vm
+                .insert_value(block_number_addr, Felt252::from(block.number))
+                .map_err(HintError::Memory);
+
+            // Insert the block gas limit into the VM.
+            let _ = vm
+                .insert_value(block_gas_limit_addr, Felt252::from(block.gas_limit))
+                .map_err(HintError::Memory);
+
+            // Insert the block timestamp into the VM.
+            let _ = vm
+                .insert_value(block_timestamp_addr, Felt252::from(block.timestamp))
+                .map_err(HintError::Memory);
+
+            // Insert the coinbase (beneficiary) address into the VM.
+            let _ = vm
+                .insert_value(coinbase_addr, Felt252::from_bytes_be_slice(&block.beneficiary.0 .0))
+                .map_err(HintError::Memory);
+
+            // Insert the base fee of the block into the VM, defaulting to 0 if not set.
+            let _ = vm
+                .insert_value(
+                    base_fee_addr,
+                    Felt252::from(block.base_fee_per_gas.unwrap_or_default()),
+                )
+                .map_err(HintError::Memory);
+
+            Ok(())
+        },
+    )
+}
+
+/// Generates a hint to add a new memory segment.
+///
+/// This function adds a hint to the `HintProcessor` that creates a new memory segment in the
+/// virtual machine. It maps the current memory pointer (`ap`) to the newly added segment.
+pub fn add_segment_hint() -> Hint {
+    Hint::new(
+        String::from("memory[ap] = to_felt_or_relocatable(segments.add())"),
+        |vm: &mut VirtualMachine,
+         _exec_scopes: &mut ExecutionScopes,
+         _ids_data: &HashMap<String, HintReference>,
+         _ap_tracking: &ApTracking,
+         _constants: &HashMap<String, Felt252>|
+         -> Result<(), HintError> {
+            // Calls the function to add a new memory segment to the VM.
+            add_segment(vm)
+        },
+    )
+}
```

Comment:
@ClementWalter here I had to do this because the hint add_segment as you wrote it in the Cairo program, is not recognized as the same code as the one in the Cairo VM Rust (https://github.com/lambdaclass/cairo-vm/blob/1f1190b5c079c777c41e35c348eaa0b5447063e8/vm/src/hint_processor/builtin_hint_processor/hint_code.rs#L1). I have the impression that there are shortcuts on the hint name on the vm in python and that it is not applicable in Rust. So either we have to change the hint code in the Cairo implementation, or we just have to add this small implementation.

Reply:
Quite surprised because this hint is used as is “everywhere” in any Cairo program (as soon as one do `alloc()`. Would it be due to compiler version ? We definitishouldnt miss a hint for such a basic operation

Reply:
Me too was very surprised but if I remove this I've this error:
<img width="781" alt="image" src="https://github.com/user-attachments/assets/44f7b5cc-a8c2-49d3-b3e1-ef68d5797db7">

I don't really understand why the hint code is not interpreted the same way but the most likely hypothesis seems to me to be that there is a simplification transformation that is done in the Cairo VM python to simplify the writing of the hint while it is not found in Rust?

I also have the impression that the nondet modifies the hint code, I don't know why? I tried to search in the python code but I didn't find anything clear explaining this... Maybe we can merge as is and will track this in a followup PR to eliminate this, wdyt?

Reply:
so you're right, using `nondet` doesn't put the hint as is but replace it with what you've got. This hint is consequently also implemented in the https://github.com/keep-starknet-strange/snos/blob/d2248cd7df5ed565bea2781a639a45c88042f254/crates/starknet-os/src/hints/mod.rs#L508